File: 08_SI/01_SimpleDms/SimpleDms/src/main/java/org/example/simpledms/service/basic/EmpService.java
package org.example.simpledms.service.basic;

import org.example.simpledms.model.entity.basic.Emp;
import org.example.simpledms.repository.basic.EmpRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * packageName : org.example.jpaexam.service.basic
 * fileName : EmpService
 * author : GGG
 * date : 2024-03-20
 * description : 사원 서비스 : DB JPA CRUD 함수이용 업무로직 만드는 곳
 * 요약 :
 * <p>
 * ===========================================================
 * DATE            AUTHOR             NOTE
 * -----------------------------------------------------------
 * 2024-03-20         GGG          최초 생성
 */
@Service
public class EmpService {

    //    DB JPA CRUD 클래스 받기 : DI(의존성 주입)
//         DI(Dependency(의존성) Injection(주입))
    @Autowired
    EmpRepository empRepository;

    //    전체조회
    public List<Emp> findAll() {
//        DB JPA 전체조회 함수 호출 : JPA 자동 sql 문 작성함
        List<Emp> list = empRepository.findAll();
        return list;
    }

    //    전체조회 + like 검색 함수
    public Page<Emp> findAllByEnameContaining(
            String ename,
            Pageable pageable
    ) {
        Page<Emp> page
                = empRepository
                .findAllByEnameContaining(ename, pageable);
        return page;
    }

    //    TODO: 연습 : EmpService 클래스에서 상세조회 findById()
//                EmpController 클래스에서 getEmpId()
//                emp_id.jsp(테스트목적)
    public Optional<Emp> findById(int eno) {
//        DB 상세조회 함수 : JPA 상세조회함수 : findById
        Optional<Emp> optionalEmp
                = empRepository.findById(eno);
        return optionalEmp;
    }

    //    TODO: 연습 : EmpService 클래스에서 저장/수정 save()
//                EmpController 클래스에서 addEmp()
//                add_emp.jsp
//      url : /emp/addition
    public Emp save(Emp emp) {
//        JPA 기본 저장함수 실행 : save()
        Emp emp2 = empRepository.save(emp);

        return emp2;
    }

    //   TODO: 연습 : EmpService.java => 삭제함수 removeById()
//               EmpController.java => deleteEmp()
//               jsp : 수정페이지에 삭제버튼 추가
    public boolean removeById(int eno) {
        if (empRepository.existsById(eno) == true) {
            empRepository.deleteById(eno); // DB 삭제
            return true;
        } else {
            return false;
        }

    }
}

File: 08_SI/05_Login/frontend-vue/src/services/basic/EmpService.js
// EmpService.js
import http from "@/utils/http-common";
import AuthHeader from "../auth/AuthHeader";

class EmpService {
  // TODO: 전체 조회
  getAll(ename, page, size) {
    return http.get(`/basic/emp?ename=${ename}&page=${page}&size=${size}`,{
      headers: AuthHeader()
    });
  }

  create(data) {
    return http.post("/basic/emp", data,{
      headers: AuthHeader()
    });
  }

  get(eno) {
    return http.get(`/basic/emp/${eno}`,{
      headers: AuthHeader()
    });
  }

  // TODO: 수정함수 : 사원번호(eno), 수정할객체(data)
  update(eno, data) {
    return http.put(`/basic/emp/${eno}`, data,{
      headers: AuthHeader()
    });
  }

  // TODO: 삭제함수 : 사원번호(eno)
  delete(eno) {
    return http.delete(`/basic/emp/deletion/${eno}`,{
      headers: AuthHeader()
    });
  }
}

export default new EmpService();